runtime/domain.c

```diff
@@ -2214,7 +2205,7 @@ bool caml_free_domains(void)
     struct dom_internal* dom = &all_domains[i];
 
     /* Give the backup thread time to terminate gracefully, if needed */
-    while (atomic_load_acquire(&dom->backup_thread_running) != 0) {
+    while (backup_thread_running(dom)) {
       cpu_relax();
     }
```

Comment:
When this particular code path is hit, we are in "cleanup at exit" mode, and want to ensure, while terminating a domain, that its backup thread terminates, before processing the next domain. In this case the backup thread has already been properly signaled, what really is needed is for it to be scheduled - either on the current processor or on another one. Using `cpu_relax` here should be acceptable because the loop really won't take long.

Reply:
I'm not convinced (I would rather have a proper synchronization operation within the busy loop), but then I realized that we cannot do what the other loop does anyway, it releases and retakes the domain lock but here we are trying to terminate the backup thread of *other* domains whose domain lock we do not own. I'm leaving the code as-is.